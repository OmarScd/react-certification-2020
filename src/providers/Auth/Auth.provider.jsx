import React, { useState, useEffect, useContext } from 'react';

import { USERNAME_STORAGE_KEY } from '../../utils/constants';
import { storage } from '../../utils/storage';

const AuthContext = React.createContext(null);

const useAuth = () => {
  const context = useContext(AuthContext);
  if (!context) {
    throw new Error(`Can't use "useAuth" without an AuthProvider!`);
  }
  return context;
};

const AuthProvider = ({ children }) => {
  const [authenticated, setAuthenticated] = useState(false);
  const [username, setUsername] = useState(undefined);

  useEffect(() => {
    const user = storage.get(USERNAME_STORAGE_KEY);
    const isAuthenticated = !!user;

    setAuthenticated(isAuthenticated);
    if (isAuthenticated) {
      setUsername(user);
    }
  }, []);

  const login = (name) => {
    setAuthenticated(true);
    setUsername(name);
    storage.set(USERNAME_STORAGE_KEY, name);
  };

  const logout = () => {
    setAuthenticated(false);
    setUsername(undefined);
    storage.remove(USERNAME_STORAGE_KEY);
  };

  return (
    <AuthContext.Provider value={{ login, logout, authenticated, username }}>
      {children}
    </AuthContext.Provider>
  );
};

export { useAuth, AuthProvider as default };
