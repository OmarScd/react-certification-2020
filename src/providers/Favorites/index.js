export { default } from './Favorites.provider';
