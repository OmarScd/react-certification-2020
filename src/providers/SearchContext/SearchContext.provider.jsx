import React, { useState, useContext } from 'react';

const SearchContext = React.createContext();

const useSearch = () => {
  const context = useContext(SearchContext);
  if (!context) {
    throw new Error(`Can't use "useSearch" without a FavoritesProvider!`);
  }
  return context;
};

const SearchContextProvider = ({ children }) => {
  const [query, setQuery] = useState('Subdivisions');

  return (
    <SearchContext.Provider value={{ query, setQuery }}>
      {children}
    </SearchContext.Provider>
  );
};

export { useSearch, SearchContextProvider as default };
