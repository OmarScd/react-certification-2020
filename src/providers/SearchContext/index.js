export { default, useSearch } from './SearchContext.provider';
