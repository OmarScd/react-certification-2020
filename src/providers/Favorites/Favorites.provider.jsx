/* eslint-disable react-hooks/exhaustive-deps */
import React, { useEffect, useContext, useReducer } from 'react';

import { USERNAME_STORAGE_KEY } from '../../utils/constants';
import { storage } from '../../utils/storage';
import { useAuth } from '../Auth';
import favoritesReducer from '../../reducers/favorites';

const FavoritesContext = React.createContext([]);

const useFavorites = () => {
  const context = useContext(FavoritesContext);
  if (!context) {
    throw new Error(`Can't use "useFavorites" without a FavoritesProvider!`);
  }
  return context;
};

const FavoritesProvider = ({ children }) => {
  const [favorites, favoritesDispatch] = useReducer(favoritesReducer, []);
  const { username, authenticated } = useAuth();
  const userFavoritesStorageKey = `${USERNAME_STORAGE_KEY}-${username}`;

  useEffect(() => {
    if (authenticated) {
      const favoritesData = storage.get(userFavoritesStorageKey);
      if (favoritesData) {
        favoritesDispatch({ type: 'POPULATE_FAVORITES', favorites: favoritesData });
      }
    } else {
      favoritesDispatch({ type: 'CLEAR_FAVORITES' });
    }
  }, [authenticated]);

  useEffect(() => {
    if (authenticated) {
      storage.set(userFavoritesStorageKey, favorites);
    }
  }, [favorites]);

  return (
    <FavoritesContext.Provider value={{ favorites, favoritesDispatch }}>
      {children}
    </FavoritesContext.Provider>
  );
};

export { useFavorites, FavoritesProvider as default };
