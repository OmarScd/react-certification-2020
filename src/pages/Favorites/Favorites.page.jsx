import React from 'react';
import styled from 'styled-components';

import Layout from '../../components/Layout';
import VideoCard from '../../components/VideoCard';
import { useFavorites } from '../../providers/Favorites';

const Container = styled.div`
  display: flex;
  flex-wrap: wrap;
  margin: 0 auto;
  max-width: 1500px;
  padding: 0 3rem;
  justify-content: center;
`;

const FavoritesPage = () => {
  const { favorites } = useFavorites();

  return (
    <Layout>
      <Container>
        {favorites.length > 0 ? (
          favorites.map((item) => (
            <VideoCard
              key={item.id}
              id={item.id}
              thumbnail={item.thumbnail}
              title={item.title}
              description={item.description}
            />
          ))
        ) : (
          <p>You don&apos;t have any favorites yet</p>
        )}
      </Container>
    </Layout>
  );
};

export default FavoritesPage;
