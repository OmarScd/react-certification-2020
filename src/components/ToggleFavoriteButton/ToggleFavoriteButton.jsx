import React, { useState, useEffect } from 'react';
import styled from 'styled-components';

import { useFavorites } from '../../providers/Favorites';
import { useAuth } from '../../providers/Auth';

const Button = styled.button`
  text-transform: uppercase;
`;

const ToggleFavoriteButton = ({ id, thumbnail, title, description }) => {
  const { favorites, favoritesDispatch } = useFavorites();
  const { authenticated } = useAuth();

  const [isFavorite, setIsFavorite] = useState(false);

  useEffect(() => {
    setIsFavorite(favorites.findIndex((favorite) => favorite.id === id) !== -1);
  }, [favorites, id]);

  const handleOnClick = () => {
    if (isFavorite) {
      favoritesDispatch({
        type: 'REMOVE_FAVORITE',
        id,
      });
    } else {
      favoritesDispatch({
        type: 'ADD_FAVORITE',
        favorite: {
          id,
          thumbnail,
          title,
          description,
        },
      });
    }
  };

  return (
    <>
      {authenticated ? (
        <Button onClick={handleOnClick}>
          {isFavorite ? 'Remove from' : 'Add to'} favorites
        </Button>
      ) : (
        <></>
      )}
    </>
  );
};

export default ToggleFavoriteButton;
