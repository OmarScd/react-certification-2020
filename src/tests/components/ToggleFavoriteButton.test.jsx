import React from 'react';
import { render, cleanup } from '@testing-library/react';
import ToggleFavoriteButton from '../../components/ToggleFavoriteButton';
import ThemeProvider from '../../providers/Theme';
import '../../providers/SearchContext';
import '../../providers/Auth';
import '../../providers/Favorites';

jest.mock('../../providers/SearchContext', () => ({
  default: jest.fn(),
  useSearch: () => ({ query: 'test', setQuery: jest.fn() }),
}));

jest.mock('../../providers/Auth', () => ({
  default: jest.fn(),
  useAuth: () => ({ authenticated: true }),
}));

jest.mock('../../providers/Favorites', () => ({
  default: jest.fn(),
  useFavorites: () => {
    // eslint-disable-next-line global-require
    return {
      favorites: [
        {
          id: 'mFl8nzZuExE',
          thumbnail: 'https://i.ytimg.com/vi/mFl8nzZuExE/mqdefault.jpg',
          title: 'Shrek (2001) - Do You Know the Muffin Man? Scene (2/10) | Movieclips',
          description: 'Do You Know the Muffin Man?',
        },
      ],
      addFavorite: jest.fn(),
      removeFavorite: jest.fn(),
    };
  },
}));

const fav = {
  id: 'mFl8nzZuExE',
  thumbnail: 'https://i.ytimg.com/vi/mFl8nzZuExE/mqdefault.jpg',
  title: 'Shrek (2001) - Do You Know the Muffin Man? Scene (2/10) | Movieclips',
  description: 'Do You Know the Muffin Man?',
};

const notFav = {
  id: 'asd123',
  thumbnail: 'https://i.ytimg.com/vi/mFl8nzZuExE/mqdefault.jpg',
  title: 'Shrek (2001) - Do You Know the Muffin Man? Scene (2/10) | Movieclips',
  description: 'Do You Know the Muffin Man?',
};

describe('ToggleFavoriteButton', () => {
  afterAll(cleanup);

  it('renders correctly when element is favorite', async () => {
    const { container } = render(
      <ThemeProvider>
        <ToggleFavoriteButton {...fav} />
      </ThemeProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('renders correctly when element is not favorite', async () => {
    const { container } = render(
      <ThemeProvider>
        <ToggleFavoriteButton {...notFav} />
      </ThemeProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
