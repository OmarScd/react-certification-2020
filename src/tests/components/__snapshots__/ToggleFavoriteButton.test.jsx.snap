// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleFavoriteButton renders correctly when element is favorite 1`] = `
<div>
  <button
    class="sc-bdnylx jlICRO"
  >
    Remove from
     favorites
  </button>
</div>
`;

exports[`ToggleFavoriteButton renders correctly when element is not favorite 1`] = `
<div>
  <button
    class="sc-bdnylx jWyyAs"
  >
    Add to
     favorites
  </button>
</div>
`;
